README.md
# Runes - Simple Cookies You Can Extend (similar to Macaroons)

https://research.google/pubs/pub41892/ is a paper called "Macaroons:
Cookies with Contextual Caveats for Decentralized Authorization in the
Cloud".  It has one good idea, some extended ideas nobody implements,
and lots and lots of words.

The idea: a server issues a cookie to Alice.  She can derive cookies
with extra restrictions and hand them to Bob and Carol to send back to
the server, and they can't remove the restrictions.

But they did it using a Message Authetication Code (MAC, get it?),
which is actually counter-productive, since it's simpler and better to
use Length Extension to achieve the same results.  I call that a Rune;
this version really only handles strings, but you can use hex or another
encoding.

## Rune Language

A *rune* is a series of restrictions; you have to pass all of them (so
appending a new one always makes the rune less powerful).  Each
restriction is one or more alternatives ("cmd=foo OR cmd=bar"), any
one of which can pass.

The form of each alternative is a simple string:

    ALTERNATIVE := FIELDNAME CONDITION VALUE

`FIELDNAME` contains only UTF-8 characters, exclusive of
! " # $ % & ' ( ) * +, - . / : ;  ? @ [ \ ] ^ _ \` { | } ~ (C's ispunct()).
These can appear inside a `VALUE`, but `&`, `|` and `\\` must be escaped with `\` (escaping is legal for any character, but unnecessary).


`CONDITION` is one of the following values:
* `!`: Pass if field is missing (value ignored)
* `=`: Pass if exists and exactly equals
* `/`: Pass if exists and is not exactly equal
* `^`: Pass if exists and begins with
* `$`: Pass if exists and ends with
* `~`: Pass if exists and contains
* `<`: Pass if exists, is a valid integer (may be signed), and numerically less than
* `>`: Pass if exists, is a valid integer (may be signed), and numerically greater than
* `}`: Pass if exists and lexicograpically greater than (or longer)
* `{`: Pass if exists and lexicograpically less than (or shorter)
* `#`: Always pass: no condition, this is a comment.

Grouping using `(` and `)` may be added in future.

A restriction is a group of alternatives separated by `|`; restrictions
are separated by `&`.
e.g.

    cmd=foo | cmd=bar
	& subcmd! | subcmd{get

The first requires `cmd` be present, and to be `foo` or `bar`.  The second
requires that `subcmd` is not present, or is lexicographically less than `get`.
Both must be true for authorization to succeed.


## Rune Authorization

A run also comes with a SHA-256 authentication code.  This is
generated as SHA-256 of the following bytestream:

1. The secret (less than 56 bytes, known only to the server which issued it).
2. For every restriction:
   1. Pad the stream as per SHA-256 (i.e. append 0x80, then zeroes, then
      the big-endian 64-bit bitcount so far, such that it's a multiple of 64
      bytes).
   2. Append the restriction.

By using the same padding scheme as SHA-256 usually uses to end the
data, we have the property that we can initialize the SHA-256 function
with the result from any prior restriction, and continue.

The server can validate the rune authorization by repeating this
procedure and checking the result.


## Rune Encoding

Runes are encoded as base64, starting with the 256-bit SHA256
authentication code, the followed by one or more restrictions
separated by `&`.

Not because base64 is good, but because it's familiar to Web people;
we use RFC3548 with `+` and `/` replaced by `-` and `_` to make
it URL safe.

(There's also a string encoding which is easier to read and debug).

## Best Practices

It's usually worth including an id in each rune you hand out so that
you can blacklist particular runes in future (your other option is to
change your master secret, but that revokes all runes).  Because this
appears in all runes, using the empty fieldname (''), and a simple
counter reduces overall size, but you could use a UUID.

This is made trivial by the `unique_id` parameter to Rune() and
MasterRune(): it adds such an empty field with the unique id (which
the default evaluator will ignore unless you handle it explicitly).

You may also include version number, to allow future runes to have
different interpretations: this appends '-[version]' in the '' field:
the default handler will fail any cookie that has a version field
(for safe forward compatibility).

The rune unmarshalling code ensures that if an empty parameter exists,
it's the first one, and it's of a valid form.

See [examples/blacklist.py](examples/blacklist.py).


## API Example

Here's the server, making you a rune! (spoiler: it's
"-YpZTBZ4Tb5SsUz3XIukxBxR619iEthm9oNJnC0LxZM=")

```
import runes
import secrets

# Secret determined by fair dice roll.
secret = bytes([5] * 16)

# Make an unrestricted rune.
rune = runes.MasterRune(secret)

# We could add our own restrictions here, if we wanted.
print("Your rune is {}".format(rune.to_base64()))
```

Here's the server, checking a rune.  You will need to define what
conditions you provide for the rune to test; one of the most useful
ones is time, but other common things are the resource being accessed,
(e.g. URL, or command and parameters), or who is accessing it (assuming
you have authenticated them already in some way).

```
import runes
import time
import sys

secret = bytes([5] * 16)

# In real life, this would come from the web data.
runestring = sys.argv[1]

# This checks the format is correct, it's authorized, an that it meets
# our values.  I assume we have values time (UNIX, seconds since
# 1970), command and optional id.
# (You can also use rune.check() if you don't care *why* it failed)
ok, whyfail = rune.check_with_reason(secret, runestring,
                                     {'time': int(time.time()),
                                      'command': 'somecommand',
                                      'id': 'DEADBEEF'})
if not ok:
    print("Rune restrictions failed: {}".format(whyfail))
    sys.exit(1)

print("Yes, you passed!")
```


Here's the client Alice.  She gets the rune and gives Bob a variant
that can only be used for 1 minute:

```
import runes
import time

# In real life, this would come from the web data.
runestring = sys.argv[1]

# You'd catch exceptions here, usually.
rune = runes.from_base64(runestring)

# You can construct a Restriction class from a sequence of Alternative
# but it's easier to use decode() to translate a string
rune.add_restriction(rune.Restriction.decode("time < {}".format((int)time.time() + 60))

print("Your restricted rune is {}".format(rune.to_base64()))
```

You can find more examples in the examples/ subdirectory.


## Advanced Techniques

If you place a callable in the dictionary to check(), that will be
called if referred to by a restriction, so you can perform your own
processing.

This is useful in implementing ratelimiting, for example: you can have
a last-used time for each id, and thus fail if it is too soon.

See [examples/ratelimit.py](examples/ratelimit.py).


## Author

Rusty Russell wrote it; but I blame @roasbeef for raving about them
long enough at LnConf that I actually read the paper.  It only took me
18 months to find a day to implement them.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.5.0] - 2022-06-22

## Changed
 - Rune() API change: unique_id comes first, and constructor generates a valid Rune from a master authcode() as expected.
 - stricter checks on use of unique_id field to avoid caller having to.

### Added
 - to_str()/from_str() API for encoding Runes without base64.
 - tests/ now contains test vectors, for other implementations.

## [0.4.0] - 2021-08-11

## Changed
 - empty field without hypen ignored by default (unless in values dict)

### Added
 - Callables in dict allow custom evaluation
 - id field parameters added to constructors for convenience
 - New examples for blacklisting and ratelimiting.

## [0.3.1] - 2021-08-02

## Fixed
 - copy() now correctly copies (original not modified when restrictions added!)

## [0.3] - 2021-08-01

## Changed
 - Punctuation is now allowed in field values; \ is the escape mechanism

### Added
 - Not equals operator (/) added.

## [0.2] - 2021-07-31

### Added

 - examples/decode.py: pretty print output into english phrases
 - Rune.copy() helper, and copy utils now work (eg. copy.deepcopy(Rune))
 - Simple check() and check_with_reason() routines for simplest usage.

## [0.1] - 2021-07-30

Initial release.


runes/runes.py
import base64
import copy
import hashlib
import re
# We can't use the hashlib one, since we need midstate access :(
import sha256  # type: ignore
import string
from typing import Dict, List, Sequence, Optional, Tuple, Any, Union


def padlen_64(x: int):
    """Amount which will increase x until it's divisible evenly by 64"""
    return (64 - (x % 64)) % 64


def end_shastream(length: int):
    """Simulate an SHA-256 ending pad (1 bit, zero bad, 64-bit length)"""
    padlen = padlen_64(length + 1 + 8)
    return bytes([0x80]) + bytes(padlen) + int.to_bytes(length * 8, 8, 'big')


class Alternative(object):
    """One of possibly several conditions which could be met"""
    def __init__(self, field: str, cond: str, value: str, allow_idfield: bool = False):
        if any([c in string.punctuation for c in field]):
            raise ValueError("field not valid")
        if cond not in ('!', '=', '/', '^', '$', '~', '<', '>', '}', '{', '#'):
            raise ValueError("cond not valid")
        if field == '':
            if not allow_idfield:
                raise ValueError("unique_id field not valid here")
            if cond != '=':
                raise ValueError("unique_id condition must be '='")
        self.field = field
        self.value = value
        self.cond = cond

    def is_unique_id(self) -> bool:
        return self.field == ''

    def test(self, values: Dict[str, Any]) -> Optional[str]:
        """Returns None on success, otherwise an explanation string"""
        # This is always True
        if self.cond == '#':
            return None

        def why(cond, field, explanation) -> Optional[str]:
            if cond:
                return None
            return '{}: {}'.format(field, explanation)

        # If it's missing, it's only True if it's a missing test.
        if self.field not in values:
            # Default to ignoring id field as long as no version.
            if self.is_unique_id():
                return why('-' not in self.value, 'id', 'unknown version {}'.format(self.value))
            return why(self.cond == '!', self.field, 'is missing')

        # If they supply a function, hand it to them.
        if callable(values[self.field]):
            return values[self.field](self)

        val = str(values[self.field])
        if self.cond == '!':
            return why(False, self.field, 'is present')
        elif self.cond == '=':
            return why(val == self.value,
                       self.field,
                       '!= {}'.format(self.value))
        elif self.cond == '/':
            return why(val != self.value,
                       self.field,
                       '= {}'.format(self.value))
        elif self.cond == '^':
            return why(val.startswith(self.value),
                       self.field,
                       'does not start with {}'.format(self.value))
        elif self.cond == '$':
            return why(val.endswith(self.value),
                       self.field,
                       'does not end with {}'.format(self.value))
        elif self.cond == '~':
            return why(self.value in val,
                       self.field,
                       'does not contain {}'.format(self.value))
        elif self.cond == '<':
            try:
                actual_int = int(val)
            except ValueError:
                return why(False, self.field, "not an integer field")
            try:
                restriction_val = int(self.value)
            except ValueError:
                return why(False, self.field, "not a valid integer")
            return why(actual_int < restriction_val,
                       self.field,
                       ">= {}".format(restriction_val))
        elif self.cond == '>':
            try:
                actual_int = int(val)
            except ValueError:
                return why(False, self.field, "not an integer field")
            try:
                restriction_val = int(self.value)
            except ValueError:
                return why(False, self.field, "not a valid integer")
            return why(actual_int > restriction_val,
                       self.field,
                       "<= {}".format(restriction_val))
        elif self.cond == '{':
            return why(val < self.value,
                       self.field,
                       'is the same or ordered after {}'.format(self.value))
        elif self.cond == '}':
            return why(val > self.value,
                       self.field,
                       'is the same or ordered before {}'.format(self.value))
        else:
            # We checked this in init!
            assert False

    def encode(self) -> str:
        return self.field + self.cond + (self.value
                                         .replace('\\', '\\\\')
                                         .replace('|', '\\|')
                                         .replace('&', '\\&'))

    @classmethod
    def decode(cls, encstr: str, allow_idfield: bool = False) -> Tuple['Alternative', str]:
        """Pull an Alternative from encoded string, return remainder"""
        cond = None
        end_off = 0

        # Swallow field up to conditiona
        while end_off < len(encstr):
            if encstr[end_off] in string.punctuation:
                cond = encstr[end_off]
                break
            end_off += 1
        if cond is None:
            raise ValueError('{} does not contain any operator'
                             .format(encstr))
        field = encstr[:end_off]
        end_off += 1

        value = ''
        while end_off < len(encstr):
            if encstr[end_off] == '|':
                # We swallow this
                end_off += 1
                break
            if encstr[end_off] == '&':
                break
            if encstr[end_off] == '\\':
                end_off += 1
            value += encstr[end_off]
            end_off += 1

        return cls(field, cond, value, allow_idfield), encstr[end_off:]

    @classmethod
    def from_str(cls, encstr: str) -> 'Alternative':
        """Turns this user-readable string into an Alternative (no escaping)"""
        encstr = re.sub(r'\s+', '', encstr)
        parts = re.split('([' + string.punctuation + '])', encstr, maxsplit=1)
        return cls(parts[0], parts[1], parts[2])

    def __eq__(self, other) -> bool:
        return (self.field == other.field
                and self.value == other.value
                and self.cond == other.cond)


class Restriction(object):
    """A restriction is a set of alternatives: any of those pass, the
restriction is met"""
    def __init__(self, alternatives: Sequence[Alternative]):
        if alternatives == []:
            raise ValueError("Restriction must have some alternatives")
        self.alternatives = alternatives

    def test(self, values: Dict[str, Any]) -> Optional[str]:
        """Returns None on success, otherwise a string of all the failures"""
        reasons = []
        for alt in self.alternatives:
            reason = alt.test(values)
            if reason is None:
                return None
            reasons.append(reason)

        return " AND ".join(reasons)

    def encode(self) -> str:
        return '|'.join([alt.encode() for alt in self.alternatives])

    @classmethod
    def decode(cls, encstr: str, allow_idfield: bool = False) -> Tuple['Restriction', str]:
        """Pull a Restriction from encoded string, return remainder"""
        alts = []
        while len(encstr) != 0:
            if encstr.startswith('&'):
                encstr = encstr[1:]
                break
            alt, encstr = Alternative.decode(encstr, allow_idfield)
            alts.append(alt)
            # We never allow id fields after first.
            allow_idfield = False

        if len(alts) > 1 and alts[0].is_unique_id():
            raise ValueError("unique_id field cannot have alternatives")

        return cls(alts), encstr

    @classmethod
    def from_str(cls, encstr: str) -> 'Restriction':
        """Returns a Restriction from an escaped string (ignoring whitespace)"""
        encstr = re.sub(r'\s+', '', encstr)
        ret, remainder = cls.decode(encstr)
        if len(remainder) != 0:
            raise ValueError("Restriction had extrs characters at end: {}"
                             .format(remainder))
        return ret

    @classmethod
    def unique_id(cls,
                  unique_id: Union[int, str],
                  version: Optional[Union[int, str]] = None) -> 'Restriction':
        """Helper to produce an id 'restriction'"""
        idstr = str(unique_id)
        if '-' in idstr:
            raise ValueError('Hyphen not allowed in unique_id {}'.format(idstr))
        if version:
            idstr += '-{}'.format(version)
        # We use the empty field for this, since it's always present.
        return cls([Alternative('', '=', idstr, allow_idfield=True)])

    def __eq__(self, other) -> bool:
        return list(self.alternatives) == list(other.alternatives)


class Rune(object):
    """A Rune, such as you might get from a server.  You can add
restrictions and it will still be valid"""
    def __init__(self,
                 authbase: bytes,
                 unique_id: Optional[Union[int, str]] = None,
                 version: Optional[Union[int, str]] = None,
                 restrictions: Sequence[Restriction] = []):
        """Convenient constructor: adds unique_id, version and any other restrictions"""
        assert isinstance(unique_id, (int, str, type(None)))
        assert isinstance(version, (int, str, type(None)))
        self.restrictions: List[Restriction] = []

        # Replace with real shastate (aka authcode)
        self.shaobj = sha256.sha256()
        self.shaobj.state = (authbase, 64)

        # If they provide a unique_id, it goes first.
        if unique_id is not None:
            self.add_restriction(Restriction.unique_id(unique_id, version))

        for r in restrictions:
            self.add_restriction(r)

    @classmethod
    def from_authcode(cls,
                      authcode: bytes,
                      restrictions: Sequence[Restriction]):
        """Alternate constructor when authcode already calculated with these restrictions"""
        ret = cls(authcode)
        ret.restrictions = list(restrictions)

        # SHA state needs to simply be updated to cover this length.
        runelength = ret.shaobj.state[1]
        for r in ret.restrictions:
            runelength += len(r.encode())
            runelength += padlen_64(runelength)

        ret.shaobj.state = (ret.shaobj.state[0], runelength)
        return ret

    def add_restriction(self, restriction: Restriction) -> None:
        self.restrictions.append(restriction)
        self.shaobj.update(bytes(restriction.encode(), encoding='utf8'))
        self.shaobj.update(end_shastream(self.shaobj.state[1]))

    def are_restrictions_met(self, values: Dict[str, Any]) -> Tuple[bool, str]:
        """Tests the restrictions against the values dict given.  Normally
        values are treated strings, but < and > conditions only work
        if they're actually integers.

        Returns (True, '') if everything is good.  Otherwise, returns
        (False, reasonstring)

        """,
        for r in self.restrictions:
            reasons = r.test(values)
            if reasons is not None:
                return False, reasons
        return True, ''

    def authcode(self) -> bytes:
        return self.shaobj.state[0]

    def to_str(self) -> str:
        return (self.authcode().hex()
                + ':'
                + '&'.join([r.encode() for r in self.restrictions]))

    def to_base64(self) -> str:
        restrstr = '&'.join([r.encode() for r in self.restrictions])
        binstr = base64.urlsafe_b64encode(self.authcode()
                                          + bytes(restrstr, encoding='utf8'))
        return binstr.decode('utf8')

    @classmethod
    def from_str(cls, rstr: str) -> 'Rune':
        if len(rstr) < 64 or rstr[64] != ':':
            raise ValueError("Rune strings must start with 64 hex digits then '-'")
        authcode = bytes.fromhex(rstr[:64])
        restrictions: List[Restriction] = []
        restrictstr = rstr[65:]

        while len(restrictstr) != 0:
            # ID field is only valid at front!
            allow_idfield = (restrictions == [])
            restr, restrictstr = Restriction.decode(restrictstr,
                                                    allow_idfield=allow_idfield)
            restrictions.append(restr)
        return cls.from_authcode(authcode, restrictions)

    @classmethod
    def from_base64(cls, b64str: str) -> 'Rune':
        binstr = base64.urlsafe_b64decode(b64str)
        return cls.from_str(binstr[:32].hex() + ':' + binstr[32:].decode('utf8'))

    def __eq__(self, other) -> bool:
        return (self.restrictions == other.restrictions
                and self.shaobj.state == other.shaobj.state)

    def copy(self) -> 'Rune':
        """Perform a shallow copy"""
        return self.__copy__()

    def __copy__(self) -> 'Rune':
        # You don't want to share the shaobj!
        return self.from_authcode(self.shaobj.state[0], self.restrictions)

    def __deepcopy__(self, memo=None) -> 'Rune':
        """sha256.sha256 doesn't implement pickle"""
        return self.from_authcode(self.shaobj.state[0], copy.deepcopy(self.restrictions))


class MasterRune(Rune):
    """This is where the server creates the Rune; it's recommended you
give each rune a unique id (often a persistent counter) (with an
optional version), which gets included as an empty-fieldname field.

    """
    def __init__(self,
                 seedsecret: bytes,
                 restrictions: Sequence[Restriction] = [],
                 unique_id: Optional[Union[int, str]] = None,
                 version: Optional[Union[int, str]] = None):
        # If they provide a unique_id, it goes first.
        if unique_id is not None:
            restrictions = [Restriction.unique_id(unique_id, version)] + list(restrictions)

        self.restrictions = []
        # Everyone assumes that seed secret takes 1 block only
        assert len(seedsecret) + 1 + 8 <= 64
        self.shaobj = sha256.sha256()
        self.shaobj.update(seedsecret + end_shastream(len(seedsecret)))
        for r in restrictions:
            self.add_restriction(r)

        # For fast calc using hashlib
        self.shabase = hashlib.sha256()
        self.shabase.update(seedsecret)
        self.seclen = len(seedsecret)

    def copy(self) -> 'Rune':
        """Perform a shallow copy"""
        return self.__copy__()

    def __copy__(self) -> 'MasterRune':
        # Create dummy so we can populate it (we don't store secret)
        ret = MasterRune(bytes())
        ret.restrictions = self.restrictions.copy()
        ret.shaobj.state = self.shaobj.state
        ret.shabase = self.shabase
        ret.seclen = self.seclen
        return ret

    def __deepcopy__(self, memo=None) -> 'MasterRune':
        """sha256.sha256 doesn't implement pickle"""
        ret = MasterRune(bytes())
        ret.restrictions = copy.deepcopy(self.restrictions)
        ret.shaobj.state = self.shaobj.state
        ret.shabase = self.shabase
        ret.seclen = self.seclen
        return ret

    def is_rune_authorized(self, other: Rune) -> bool:
        """This is faster than adding the restrictions one-by-one and checking
        the final authcode (but equivalent)"""
        # Make copy, as we're going to update state.
        sha = self.shabase.copy()
        totlen = self.seclen
        for r in other.restrictions:
            pad = end_shastream(totlen)
            sha.update(pad)
            totlen += len(pad)
            enc = bytes(r.encode(), encoding='utf8')
            sha.update(enc)
            totlen += len(enc)

        return other.authcode() == sha.digest()

    def check_with_reason(self, b64str: str, values: Dict[str, Any]) -> Tuple[bool, str]:
        """All-in-one check that a runestring is valid, derives from this
MasterRune and passes all its conditions against the given dictionary
of values or callables"""
        try:
            rune = Rune.from_base64(b64str)
        except:  # noqa: E722
            return False, "runestring invalid"
        if not self.is_rune_authorized(rune):
            return False, "rune authcode invalid"
        return rune.are_restrictions_met(values)


def check_with_reason(secret: bytes, b64str: str, values: Dict[str, Any]) -> Tuple[bool, str]:
    """Convenience function that the b64str runestring is valid, derives
from our secret, and passes against these values.  If you want to
check many runes, it's more efficient to create the MasterRune first
then check them, but this is fine if you're only checking one.

    """
    return MasterRune(secret).check_with_reason(b64str, values)


def check(secret: bytes, b64str: str, values: Dict[str, Any]) -> bool:
    """Convenience function that the b64str runestring is valid, derives
from our secret, and passes against these values.  If you want to
check many runes, it's more efficient to create the MasterRune first
then check them, but this is fine if you're only checking one.

Unlike check_with_reason(), this discards the reason and returns a
simple True or False.

    """
    return check_with_reason(secret, b64str, values)[0]
